pandas/core/reshape/tile.py

```diff
@@ -340,6 +340,9 @@ def qcut(
     x = _preprocess_for_cut(x)
     x, dtype = _coerce_to_type(x)
 
+    if is_extension_array_dtype(x.dtype) and is_integer_dtype(x.dtype):
```

Comment:
this needs to go in _coerce_to_type

Reply:
So if I put this into _coerce_to_type the qcut tests pass but not the cut tests, because it also changes behavior here: https://github.com/pandas-dev/pandas/blob/master/pandas/core/reshape/tile.py#L208. @TomAugspurger might be able to shed some light on why x needs to be object dtype in cut?

Reply:
move this code: https://github.com/pandas-dev/pandas/blob/master/pandas/core/reshape/tile.py#L210

inside _coerce_to_type and might fix all

Reply:
This does seem to work, but only if we modify the cut test to operate on float arrays instead of object arrays with NaN. Seems something odd might be happening with cut (this is master):

```python
In [5]: a = np.arange(100).astype(object)                                                                                                                                       

In [6]: a[::2] = np.nan                                                                                                                                                         

In [7]: a                                                                                                                                                                       
Out[7]: 
array([nan, 1, nan, 3, nan, 5, nan, 7, nan, 9, nan, 11, nan, 13, nan, 15,
       nan, 17, nan, 19, nan, 21, nan, 23, nan, 25, nan, 27, nan, 29, nan,
       31, nan, 33, nan, 35, nan, 37, nan, 39, nan, 41, nan, 43, nan, 45,
       nan, 47, nan, 49, nan, 51, nan, 53, nan, 55, nan, 57, nan, 59, nan,
       61, nan, 63, nan, 65, nan, 67, nan, 69, nan, 71, nan, 73, nan, 75,
       nan, 77, nan, 79, nan, 81, nan, 83, nan, 85, nan, 87, nan, 89, nan,
       91, nan, 93, nan, 95, nan, 97, nan, 99], dtype=object)

In [8]: object_result = pd.cut(a, 3)                                                                                                                                            

In [9]: float_result = pd.cut(a.astype(float), 3)                                                                                                                               

In [10]: assert_categorical_equal(object_result, float_result)   
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-10-4469237ad337> in <module>
----> 1 assert_categorical_equal(object_result, float_result)

~/pandas/pandas/_testing.py in assert_categorical_equal(left, right, check_dtype, check_category_order, obj)
    843         assert_index_equal(left.categories, right.categories, obj=f"{obj}.categories")
    844         assert_numpy_array_equal(
--> 845             left.codes, right.codes, check_dtype=check_dtype, obj=f"{obj}.codes",
    846         )
    847     else:

~/pandas/pandas/_testing.py in assert_numpy_array_equal(left, right, strict_nan, check_dtype, err_msg, check_same, obj)
   1000     # compare shape and values
   1001     if not array_equivalent(left, right, strict_nan=strict_nan):
-> 1002         _raise(left, right, err_msg)
   1003 
   1004     if check_dtype:

~/pandas/pandas/_testing.py in _raise(left, right, err_msg)
    994             diff = diff * 100.0 / left.size
    995             msg = f"{obj} values are different ({np.round(diff, 5)} %)"
--> 996             raise_assert_detail(obj, msg, left, right)
    997 
    998         raise AssertionError(err_msg)

~/pandas/pandas/_testing.py in raise_assert_detail(obj, message, left, right, diff)
    927         msg += f"\n[diff]: {diff}"
    928 
--> 929     raise AssertionError(msg)
    930 
    931 

AssertionError: Categorical.codes are different

Categorical.codes values are different (33.0 %)
[left]:  [-1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0]
[right]: [-1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2]
```

Looks like the result for object dtype is off:

```python
In [16]: a[-5:]                                                                                                                                                                 
Out[16]: array([95, nan, 97, nan, 99], dtype=object)

In [17]: object_result[-5:]                                                                                                                                                     
Out[17]: 
[(0.902, 33.667], NaN, (0.902, 33.667], NaN, (0.902, 33.667]]
Categories (3, interval[float64]): [(0.902, 33.667] < (33.667, 66.333] < (66.333, 99.0]]

In [18]: float_result[-5:]                                                                                                                                                      
Out[18]: 
[(66.333, 99.0], NaN, (66.333, 99.0], NaN, (66.333, 99.0]]
Categories (3, interval[float64]): [(0.902, 33.667] < (33.667, 66.333] < (66.333, 99.0]]
```

Without NaN we get the same output:

```python
In [21]: a = np.arange(100).astype(object)                                                                                                                                      

In [22]: object_result = pd.cut(a, 3)                                                                                                                                           

In [23]: float_result = pd.cut(a.astype(float), 3)                                                                                                                              

In [24]: assert_categorical_equal(object_result, float_result)                                                                                                                  

In [25]:       
```